virttest/utils_test/__init__.py

```diff
@@ -2318,26 +2328,179 @@ def unload_stress(stress_type, params, vms=None, remote_server=False):
                    remote_server=remote_server).unload_stress()
 
 
-def prepare_profile(test, fpath, pat_repl):
+class ServerClientStress(object):
     """
-    This is to prepare client profile to be run on client.
-    :param test: kvm test object
-    :param fpath: profile to be run
-    :param pat_repl: dict containing pattern and replacement : includes server_ip, duration, threads etc.
-    :return: no explicit return. But profile on fpath would be edited and ready to be run on client guest
-    :raise: TestError: raised if unable to edit the given profile
+    configure and run stress tools which needs server client setup
     """
-    try:
-        with open(fpath, 'r+') as profile_content:
-            tempstr = profile_content.read()
-            profile_content.truncate(0)
-            logging.debug("In prepare profile: pattern and replacement : %s", pat_repl)
-            for pattern, replace in pat_repl.items():
-                tempstr = tempstr.replace(pattern, replace)
-            profile_content.write(tempstr)
-        logging.debug("Profile xml to be run : %s ", tempstr)
-    except Exception:
-        test.error("Failed to update file : %s", fpath)
+
+    def __init__(self, test, params, env):
+        """
+        Set parameters for server client stress setup.
+        """
+
+        self.vms = env.get_all_vms()
+        self.stress_duration = int(params.get("stress_duration", "20"))
+        self.ip_rule = params.get("ip_rule", "")
+        self.stress_type = params.get("stress_type", "uperf")
+        self.need_profile = int(params.get("need_profile", 0))
+        self.server_cmd = params.get("%s_server_cmd" % self.stress_type)
+        self.client_cmd = params.get("%s_client_cmd" % self.stress_type)
+        self.client_vms = []
+        self.server_vms = []
+        if (self.need_profile):
+            self.profile = params.get("client_profile_%s" % self.stress_type)
+            self.profile = os.path.join(data_dir.get_root_dir(), self.profile)
+            self.profile_pattren = params.get("profile_pattren", "").split()
+
+        for index, vm in enumerate(self.vms):
+            if index % 2 != 0:
+                # Process for client
+                self.client_vms.append(vm)
+            else:
+                # Process for server
+                self.server_vms.append(vm)
+        if len(self.server_vms) != len(self.client_vms):
+            test.cancel(
+                "This test requires server and client vms in 1:1 ratio")
+        self.pair_vms = zip(self.server_vms, self.client_vms)
```

Comment:
I don't know the use case, just wondering if this is the common usage, some applications might have one server serve multiple clients. And the server and client sometimes might also be the host instead of vms.

Reply:
@vivianQizhu I agree you are right, and I too have same thoughts on running stress on host/remote host which I have mentioned in https://github.com/avocado-framework/avocado-vt/pull/1758#discussion_r217705163,

1. we can get no of vms to be server_vms or client_vms or host from params and default to 1:1.
2. so for windows guest we can extend implementing it in separate class.

But implementing these APIs would help us in following way which I could think of, by abstracting these stress tool configurations would help us in,

- we can run tests differently in different scenarios by tweaking the params instead of writing new testcases (reduce test code)
- reuse the existing code/params from `Stress()`/`VMStress()` and reduce redundancy in framework.

Please share your thoughts, this patch introduces client-server based stress APIs, we can work on improving and extending its functionalities. Thank you reviewing.

Reply:
> @vivianQizhu I agree you are right, and I too have same thoughts on running stress on host/remote host which I have mentioned in [#1758 (comment)](https://github.com/avocado-framework/avocado-vt/pull/1758#discussion_r217705163),
> 
>     1. we can get no of vms to be server_vms or client_vms or host from params and default to 1:1.
> 
>     2. so for windows guest we can extend implementing it in separate class.

Right, it does benefit somehow. And I agree with both of the above 2 points.

Reply:
I agree on getting no. of vms to be server_vms or client_vms from params and default to 1:1. For the host case I would be sending separate PR, since this PR is concentrated on guest side.

Reply:
done with KVM guest part